### ISSUE TITLE
How to train MLD?

### ISSUE THREAD

**Issue by serendipity800** [Xiaofeng-Tan/SoPo#2]
你好，我们有个强化学习的文章想要cite你们的文章，我们用的是MLD。
你们的文章也有MLD结果，但我看这个repo似乎只能train MDM，如何让他train MLD呢？

**Comment by Xiaofeng-Tan** [comment]
感谢您对我们工作的关注！

目前我们正在开展一项相关研究，该工作中涉及到这部分代码，因此暂时还未对外公开。我们计划在该工作完成后统一发布相关代码，预计时间在今年 3 月之后。

关于 MLD 的训练，如果你目前需要相关实现，我们有以下几点建议供参考：
- 建议关注工作 EasyTune 的开源实现 [1]，我们猜测，作者很可能会在 2 月 5 日之前公布代码；
- 你也可以自行实现 MLD 的微调。从实现角度来看，MLD 与 MDM 的主要区别在于：MLD 是在 latent space 中进行生成。因此，在 SoPo 论文中的实现思路是：先将生成结果解码，再重新编码到 reward 模型（如 TMR）中进行评估。在这一过程中，需要 冻结 VAE 的参数，仅对 diffusion 模型进行微调。除此之外，整体流程与 MDM 并无本质差异。需要注意的是，反复的编解码开销相对较小，整体计算瓶颈仍然主要来自 diffusion 过程本身；
- 在实现细节上，我们建议 使用较小的学习率，并且 谨慎筛选负样本。原因在于：负样本主要引导模型“遗忘”已有知识，而正样本则引导模型“记忆”和强化正确行为；相比之下，前者的优化难度通常显著低于后者。

如果你有进一步的问题，欢迎随时交流。我的个人联系方式可以在我的[主页](https://xiaofeng-tan.github.io/)中找到。

[1] EasyTune: Efficient Step-Aware Fine-Tuning for Diffusion-Based Motion Generation
 https://openreview.net/forum?id=Fy1EoIaAzQ

---

Thank you for your interest in our work!

At the moment, we are working on a related project that involves this part of the code, so it has not been released publicly yet. We plan to release the relevant code together with this work, which is expected around March this year.

Regarding training MLD, if you currently need an implementation, we would like to offer the following suggestions:
- We recommend keeping an eye on the open-source release of EasyTune [1]. Based on our understanding, the authors are likely to release their code before February 5;
- You may also implement MLD fine-tuning yourself. From an implementation perspective, the key difference between MLD and MDM is that MLD performs generation in the latent space. Accordingly, the implementation in the SoPo paper follows a decode–encode pipeline: the generated latent is first decoded, and then re-encoded into the reward model (e.g., TMR) for evaluation. During this process, the VAE parameters should be frozen, and only the diffusion model is fine-tuned. Apart from this, the overall pipeline is largely similar to that of MDM. Note that the computational overhead of repeated encoding and decoding is relatively small; the main bottleneck still lies in the diffusion process itself;
- For practical implementation, we suggest using a relatively small learning rate and paying close attention to negative sample selection. This is because negative samples mainly encourage the model to “forget” certain behaviors, whereas positive samples guide the model to “remember” and reinforce desirable behaviors. In practice, the former is often much easier to optimize than the latter.

If you have any further questions, please feel free to reach out. My personal contact information is available on my homepage:
👉 https://xiaofeng-tan.github.io/

Reference
[1] EasyTune: Efficient Step-Aware Fine-Tuning for Diffusion-Based Motion Generation
https://openreview.net/forum?id=Fy1EoIaAzQ

**Comment by Xiaofeng-Tan** [comment]
非常感谢您指出这个关键问题。

我解释一下当时的情况：我在整理/重构代码时，用了 AI 来帮忙做规范化和清理。但因为 README 还是沿用 MDM（Motion Diffusion Model）那套描述，AI 误以为 SOPO 相关代码是多余的，结果在清理过程中被误删了。

幸运的是，我从备份里找回了 SOPO 的原始源码，已经重新整理合并，并且调试验证跑通了。目前包含 SOPO 实现的完整代码也已经公开发布。

再次感谢你的细心提醒，也抱歉由此带来的困扰

---

Thank you very much for pointing out this critical issue.

To clarify the situation: during a recent code refactoring process, I utilized AI to improve code standardization. However, because the project's README was still based on the original MDM (Motion Diffusion Model) framework, the AI incorrectly flagged the sopo-specific modules as redundant or deprecated, leading to their accidental removal from the codebase.

Fortunately, I have retrieved the original source code from my backups. After thorough debugging and testing, I have verified that the implementation is fully functional. The repository has been updated, and the complete codebase including the SOPO implementation is now publicly available.

I appreciate your vigilance and apologize for any confusion this may have caused.

**Comment by Xiaofeng-Tan** [comment]
```
train_sopo.py (入口)
    ↓
TrainLoop.run_loop() [train/training_loop.py]
    ↓
forward_backward() [第256行]
    ↓
diffusion.training_losses(
    self.ddp_model,
    self.ref_model,        ← ref_model 被传入
    micro,
    t,
    ...
    text_model=self.text_model,
    t_m_model=self.t_m_model,
    normalizer=self.normalizer,
    ...
)
    ↓
training_losses() [gaussian_diffusion.py 第1515行] ← 🔑 SoPo 实现入口
    │
    ├─→ self.online_generate() [第1529行] 
    │   └─→ 生成两个样本并计算文本-动作相似度
    │
    ├─→ self.m_m_sim() [第1532行] 
    │   └─→ get_score_vector() [src_tmr/model/tmr.py 第30-38行] 
    │
    ├─→ self.compute_losses() [第1563-1566行] 
    │   （分别计算 model、ref_model 的损失）
    │
    └─→ 偏好优化损失计算 [第1569行] 
        loss = -log(sigmoid(-(...)))  ← DPO 风格的偏好优化损失
```

**Comment by serendipity800** [comment]
> ```
> train_sopo.py (入口)
>     ↓
> TrainLoop.run_loop() [train/training_loop.py]
>     ↓
> forward_backward() [第256行]
>     ↓
> diffusion.training_losses(
>     self.ddp_model,
>     self.ref_model,        ← ref_model 被传入
>     micro,
>     t,
>     ...
>     text_model=self.text_model,
>     t_m_model=self.t_m_model,
>     normalizer=self.normalizer,
>     ...
> )
>     ↓
> training_losses() [gaussian_diffusion.py 第1515行] ← 🔑 SoPo 实现入口
>     │
>     ├─→ self.online_generate() [第1529行] 
>     │   └─→ 生成两个样本并计算文本-动作相似度
>     │
>     ├─→ self.m_m_sim() [第1532行] 
>     │   └─→ get_score_vector() [src_tmr/model/tmr.py 第30-38行] 
>     │
>     ├─→ self.compute_losses() [第1563-1566行] 
>     │   （分别计算 model、ref_model 的损失）
>     │
>     └─→ 偏好优化损失计算 [第1569行] 
>         loss = -log(sigmoid(-(...)))  ← DPO 风格的偏好优化损失
> ```

太感谢您了！您真是十分耐心地帮我解决问题，我一定cite这个solid work
